### PR TITLE
Fix reading dirents in scalar context

### PR DESCRIPTION
**Pull request by rovo89**
Broken by 06785c554b983feb61c92ea28474fd90bae9d152 due to the extra call to nfs_readdir(). Instead of introducing another GIMME_V check, let's simplify both cases by separating the logic.

### CODE REVIEW

**Comment by rovo89** [comment]
Not sure why I didn't notice this before, but after the optimization only half of the entries were returned in scalar context.

**Comment by rovo89** [comment]
Not so much changed if you ignore whitespace: https://github.com/FGasper/p5-Net-LibNFS/pull/4/files?w=1

**Comment by FGasper** [comment]
@rovo89 Thank you for pointing this out.

I’ve pushed an alternate fix that I think is more conservative: https://github.com/FGasper/p5-Net-LibNFS/pull/5

Before I merge it, is there anything you see that I’m overlooking?

**Comment by rovo89** [comment]
> I’ve pushed an alternate fix that I think is more conservative: #5
> 
> Before I merge it, is there anything you see that I’m overlooking?

Looks ok and also a quick test worked in both scalar and list context.

Not sure if this is really "more conservative" though - I find it hard to follow the code in your version, especially for scalar context:
https://github.com/FGasper/p5-Net-LibNFS/blob/168d6a89ec62ef8c3924e85d45be1451700d9df7/LibNFS.xs#L2018-L2044

Whereas my suggestion is straightforward without any loops for the scalar version and doesn't have any breaks in the loops of the list context, so in addition to being easier to read it might also be slightly faster:
https://github.com/FGasper/p5-Net-LibNFS/blob/6bd9c2c1d42e4a556a9914c1a92547d7535174a6/LibNFS.xs#L2018-L2045

Scalar context logic is pretty much the loop of 4da187004ab93c80c234256ce95af2ec8d1d18ad unrolled.

Or, maybe even more elegant and in line with other functions, return early if no results:
```c++
struct nfsdirent* dent = nfs_readdir( perl_nfs->nfs, nfs_dh->nfsdh );

if (!dent) {
    XSRETURN_EMPTY;
}

if (GIMME_V == G_ARRAY) {
    unsigned retcount = 1;
    struct nfsdirent* cur_dent = dent;
    while (cur_dent = cur_dent->next) {
        retcount++;
    }

    EXTEND(SP, retcount);

    while (dent) {
        mPUSHs(_ptr_to_perl_dirent_obj(aTHX_ dent));
        dent = nfs_readdir( perl_nfs->nfs, nfs_dh->nfsdh );
    }

    XSRETURN(retcount);
} else {
    mXPUSHs(_ptr_to_perl_dirent_obj(aTHX_ dent));
    XSRETURN(1);
}
```
Maybe there's a way to just return `_ptr_to_perl_dirent_obj(aTHX_ dent)` in a single step, without the stack, I don't know.

Up to you of course, but please reconsider.

**Comment by FGasper** [comment]
My (gentle) preference is to minimize the duplication of logic. The unrolling has its advantages, but the duplication of the Perl object generation and stack manipulation seems to outweigh those, if only mildly.
